test/functional/pdo_sqlsrv/pdostatement_fetchmode_emulate_prepare.phpt

```diff
@@ -1,167 +1,193 @@
+        }
+    } catch (Exception $e) {
+        echo $e->getMessage() . "\n";
+    }
 
-	try {
-		$res = $conn1->setAttribute(PDO::ATTR_EMULATE_PREPARES, true);
-		if ($res)
-		{
-			echo "setAttribute should have failed.\n\n";
-		}
-	}
-	catch (Exception $e)
-	{
-		echo $e->getMessage();
-	}
+    try {
+        $query = "SELECT * FROM [$tableName]";
+        $stmt = $conn1->query($query);
+        $stmt->setAttribute(PDO::ATTR_EMULATE_PREPARES, true);
+    } catch (Exception $e) {
+        echo $e->getMessage();
+    }
 
-	echo "\n";
-	
-	try {
-		$query = "SELECT * FROM [$tableName]";
-		$stmt = $conn1->query($query);
-		$stmt->setAttribute(PDO::ATTR_EMULATE_PREPARES, true);
-	}
-	catch (Exception $e)
-	{
-		echo $e->getMessage();
-	}
+    echo "\nStart inserting data...\n";
+    $dataCols = "ID, Policy, Label";
+    $query = "INSERT INTO [$tableName](ID, Policy, Label, Budget) VALUES (?, ?, ?, ?)";
+    $stmtOptions = array(PDO::ATTR_EMULATE_PREPARES => false);
+    $stmt = $conn1->prepare($query, $stmtOptions);
+    for ($i = 1; $i <= 2; $i++) {
+        $pol = chr(64+$i);
+        $grp = "Group " . $i;
+        $budget = $i * 1000 + $i * 15;
+        $stmt->execute(array( $i, $pol, $grp, $budget ));
+    }
 
-	echo "\nStart inserting data...\n";
-	$dataCols = "ID, Policy, Label";
-	$query = "INSERT INTO [$tableName](ID, Policy, Label, Budget) VALUES (?, ?, ?, ?)";
-	$stmt = $conn1->prepare($query, array(PDO::ATTR_EMULATE_PREPARES => true));
-	$stmt = $conn1->prepare($query, array(PDO::ATTR_EMULATE_PREPARES => false));
-	for ($i = 1; $i <= 2; $i++)
-	{
-		$pol = chr(64+$i);
-		$grp = "Group " . $i;
-		$budget = $i * 1000 + $i * 15;
-		$stmt->execute( array( $i, $pol, $grp, $budget ) );
-	}
+    $query1 = "INSERT INTO [$tableName](ID, Policy, Label, Budget) VALUES (:col1, :col2, :col3, :col4)";
+    if (!isAEConnected()) {
+        $stmtOptions[PDO::ATTR_EMULATE_PREPARES] = true;
+    }
+    $stmt = $conn1->prepare($query1, $stmtOptions);
+    for ($i = 3; $i <= 5; $i++) {
+        $pol = chr(64+$i);
+        $grp = "Group " . $i;
+        $budget = $i * 1000 + $i * 15;
+        $stmt->execute(array( ':col1' => $i, ':col2' => $pol, ':col3' => $grp, ':col4' => $budget ));
+    }
+    echo "....Done....\n";
+    echo "Now selecting....\n";
+    $tsql = "SELECT * FROM [$tableName]";
+    $stmtOptions[PDO::ATTR_CURSOR] = PDO::CURSOR_SCROLL;
+    if (isColEncrypted()) {
+        $stmtOptions[PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE] = PDO::SQLSRV_CURSOR_BUFFERED;
+    }
+    $stmt1 = $conn1->prepare($tsql, $stmtOptions);
+    $stmt1->execute();
+    // The row order in the resultset when the column is encrypted (which is dependent on the encrytion key used)
+    // is different from the order when the column is not enabled
+    // To make this test work, if the column is encrypted, fetch all then find the corresponding row
+    if (!isColEncrypted()) {
+        var_dump($stmt1->fetch(PDO::FETCH_ASSOC));
```

Comment:
again, can probably choose not to encrypt the ID column so that you can enforce ORDER BY?
would it work even with emulate prepare?

Reply:
I'll skip this since they original test did not have ORDER BY.